lib/api/rule.go

```diff
@@ -26,18 +26,22 @@ import (
 type Rule struct {
 	Action string `json:"action" validate:"action"`
 
-	Protocol *Protocol `json:"protocol,omitempty" validate:"omitempty"`
-	ICMPType *int      `json:"icmpType,omitempty" validate:"omitempty,gte=0,lte=255"`
-	ICMPCode *int      `json:"icmpCode,omitempty" validate:"omitempty,gte=0,lte=255"`
+	Protocol *Protocol   `json:"protocol,omitempty" validate:"omitempty"`
+	ICMP     *ICMPFields `json:"icmp,omitempty" validate:"omitempty"`
 
-	NotProtocol *Protocol `json:"!protocol,omitempty" validate:"omitempty"`
-	NotICMPType *int      `json:"!icmpType,omitempty" validate:"omitempty,gte=0,lte=255"`
-	NotICMPCode *int      `json:"!icmpCode,omitempty" validate:"omitempty,gte=0,lte=255"`
+	NotProtocol *Protocol   `json:"!protocol,omitempty" validate:"omitempty"`
+	NotICMP     *ICMPFields `json:"!icmp,omitempty" validate:"omitempty"`
 
 	Source      EntityRule `json:"source,omitempty" validate:"omitempty"`
 	Destination EntityRule `json:"destination,omitempty" validate:"omitempty"`
 }
 
+// ICMPFields defines structure for ICMP and NotICMP sub-struct for ICMP code and type
+type ICMPFields struct {
+	Type *int `json:"type,omitempty" validate:"omitempty,gte=0,lte=255"`
```

Comment:
@robbrockbank You may still want to specify a type without a code (though not the other way around).

Since `0` is a valid ICMP code, you want it to default to `nil` so that we can distinguish "I don't care about ICMP code" from "ICMP code 0".

Type should always be present so could be made not a pointer, but changing from pointer -> int means that a default value of `0` will be picked, which is kinda unintuitive.